File: Byte_Compressor/src/core/block_codec.js
/*
 * ByteCompressor - Block-based Codec with Error Isolation v3
 *
 * Each block is independently compressed/decompressed, so a
 * transmission error in deep space corrupts only that block,
 * not the entire file (unlike ZIP/RAR).
 *
 * v3 improvements:
 *   - Larger block sizes for LZ77 profiles (less per-block overhead)
 *   - Never-expand guarantee: output is never larger than input
 *   - Optimized block header format
 */

const { crc32 } = require('./crc32');
const { compress: ansCompress, decompress: ansDecompress } = require('./context_model');
const {
    bwtForward, bwtInverse,
    mtfEncode, mtfDecode,
    deltaEncode, deltaDecode,
    rleEncode, rleDecode
} = require('./preprocessor');
const { lz77Encode, lz77Decode, lz77CompressSplit, lz77DecompressSplit } = require('./lz77');

const MAGIC = 0x4259434F; // "BYCO"
const MAGIC_RAW = 0x42594352; // "BYCR" — raw stored (never-expand mode)
const VERSION = 3;
const DEFAULT_BLOCK_SIZE = 256 * 1024;

const PROFILE = {
    GENERIC: 0,    // BWT + MTF + RLE + ANS
    TELEMETRY: 1,  // Delta + RLE + ANS
    IMAGE: 2,      // Delta + BWT + MTF + ANS
    RAW_ANS: 3,    // ANS only
    LZ77_ANS: 4,   // LZ77 + ANS (general-purpose, like improved DEFLATE)
    LZ77_TEXT: 5,   // LZ77 + BWT + MTF + ANS (text/code — best for daily use)
    LZ77_BIN: 6,    // LZ77 + RLE + ANS (binary with repeated structures)
    COUNT: 7
};

const PROFILE_NAMES = {
    [PROFILE.GENERIC]:   'GENERIC (BWT+MTF+RLE+ANS)',
    [PROFILE.TELEMETRY]: 'TELEMETRY (Delta+RLE+ANS)',
    [PROFILE.IMAGE]:     'IMAGE (Delta+BWT+MTF+ANS)',
    [PROFILE.RAW_ANS]:   'RAW_ANS (ANS only)',
    [PROFILE.LZ77_ANS]:  'LZ77+ANS (general-purpose)',
    [PROFILE.LZ77_TEXT]: 'LZ77+BWT+MTF+ANS (text/code)',
    [PROFILE.LZ77_BIN]:  'LZ77+RLE+ANS (binary)'
};

function profileName(p) { return PROFILE_NAMES[p] || 'UNKNOWN'; }

/* ── Block size selection per profile ── */

function effectiveBlockSize(profile, requestedBlockSize) {
    let blockSize = requestedBlockSize || DEFAULT_BLOCK_SIZE;

    switch (profile) {
        case PROFILE.GENERIC:
        case PROFILE.IMAGE:
            // BWT-only profiles: O(n² log n) sort, keep blocks small
            return Math.min(blockSize, 4 * 1024);

        case PROFILE.LZ77_TEXT:
            // LZ77 shrinks first, then BWT. Allow larger input blocks.
            return Math.min(blockSize, 64 * 1024);

        case PROFILE.TELEMETRY:
        case PROFILE.RAW_ANS:
        case PROFILE.LZ77_ANS:
        case PROFILE.LZ77_BIN:
            // No BWT bottleneck — use large blocks for less overhead
            return Math.min(blockSize, 512 * 1024);

        default:
            return blockSize;
    }
}

/* ── Single Block Compress ── */

function compressBlock(input, profile) {
    if (input.length === 0) return Buffer.alloc(0);

    switch (profile) {
        case PROFILE.GENERIC: {
            const { output: bwtOut, primaryIndex } = bwtForward(input);
            const mtfOut = mtfEncode(bwtOut);
            const rleOut = rleEncode(mtfOut);
            const ansOut = ansCompress(rleOut);

            const header = Buffer.alloc(4);
            header.writeInt32BE(primaryIndex, 0);
            return Buffer.concat([header, ansOut]);
        }
        case PROFILE.TELEMETRY: {
            const deltaOut = deltaEncode(input);
            const rleOut = rleEncode(deltaOut);
            return ansCompress(rleOut);
        }
        case PROFILE.IMAGE: {
            const deltaOut = deltaEncode(input);
            const { output: bwtOut, primaryIndex } = bwtForward(deltaOut);
            const mtfOut = mtfEncode(bwtOut);
            const ansOut = ansCompress(mtfOut);

            const header = Buffer.alloc(4);
            header.writeInt32BE(primaryIndex, 0);
            return Buffer.concat([header, ansOut]);
        }
        case PROFILE.RAW_ANS:
            return ansCompress(input);
        case PROFILE.LZ77_ANS: {
            // Split-stream: separate ANS per stream for better compression
            return lz77CompressSplit(input);
        }
        case PROFILE.LZ77_TEXT: {
            // Split-stream for text too (better than BWT on LZ77 output)
            return lz77CompressSplit(input);
        }
        case PROFILE.LZ77_BIN: {
            // Split-stream for binary
            return lz77CompressSplit(input);
        }
        default:
            return Buffer.alloc(0);
    }
}

/* ── Single Block Decompress ── */

function decompressBlock(input, profile, origSize) {
    if (input.length === 0) return Buffer.alloc(0);

    switch (profile) {
        case PROFILE.GENERIC: {
            const primaryIndex = input.readInt32BE(0);
            const rleData = ansDecompress(input.subarray(4));
            const mtfData = rleDecode(rleData);
            const bwtData = mtfDecode(mtfData);
            return bwtInverse(bwtData, primaryIndex);
        }
        case PROFILE.TELEMETRY: {
            const rleData = ansDecompress(input);
            const deltaData = rleDecode(rleData);
            return deltaDecode(deltaData);
        }
        case PROFILE.IMAGE: {
            const primaryIndex = input.readInt32BE(0);
            const mtfData = ansDecompress(input.subarray(4));
            const bwtData = mtfDecode(mtfData);
            const deltaData = bwtInverse(bwtData, primaryIndex);
            return deltaDecode(deltaData);
        }
        case PROFILE.RAW_ANS:
            return ansDecompress(input);
        case PROFILE.LZ77_ANS:
        case PROFILE.LZ77_TEXT:
        case PROFILE.LZ77_BIN: {
            return lz77DecompressSplit(input);
        }
        default:
            return Buffer.alloc(0);
    }
}

/* ── File-level Compress ── */

function compressFile(input, config = {}) {
    const profile = config.profile !== undefined ? config.profile : PROFILE.GENERIC;
    const blockSize = effectiveBlockSize(profile, config.blockSize);
    const blockCount = Math.ceil(input.length / blockSize);

    // File header: magic(4) + version(1) + profile(1) + blockSize(4) + origSize(4) + blockCount(4) + headerCRC(4) = 22
    const header = Buffer.alloc(22);
    let pos = 0;
    header.writeUInt32BE(MAGIC, pos); pos += 4;
    header[pos++] = VERSION;
    header[pos++] = profile;
    header.writeUInt32BE(blockSize, pos); pos += 4;
    header.writeUInt32BE(input.length, pos); pos += 4;
    header.writeUInt32BE(blockCount, pos); pos += 4;

    // Header CRC (of first 18 bytes)
    const hdrCrc = crc32(header.subarray(0, 18));
    header.writeUInt32BE(hdrCrc, 18);

    const blocks = [header];
    let totalCompressed = 22;

    for (let b = 0; b < blockCount; b++) {
        const start = b * blockSize;
        const end = Math.min(start + blockSize, input.length);
        const chunk = input.subarray(start, end);

        let compressed = compressBlock(chunk, profile);
        let usedProfile = profile;

        // If compression expands the data, store raw
        if (compressed.length === 0 || compressed.length >= chunk.length) {
            compressed = Buffer.from(chunk);
            usedProfile = 0xFF;
        }

        // Block header: origSize(4) + compSize(4) + crc(4) + profile(1) = 13
        const blkHeader = Buffer.alloc(13);
        blkHeader.writeUInt32BE(chunk.length, 0);
        blkHeader.writeUInt32BE(compressed.length, 4);
        blkHeader.writeUInt32BE(crc32(chunk), 8);
        blkHeader[12] = usedProfile;

        blocks.push(blkHeader, compressed);
        totalCompressed += 13 + compressed.length;
    }

    const result = Buffer.concat(blocks);

    // Never-expand guarantee: if compressed output >= original input,
    // return a minimal raw wrapper instead
    if (result.length >= input.length) {
        return rawStore(input);
    }

    return result;
}

/* ── Raw Store (never-expand mode) ── */

function rawStore(input) {
    // Minimal format: magic(4) + origSize(4) + crc(4) + data
    // Total overhead: 12 bytes — minimal possible
    const header = Buffer.alloc(12);
    header.writeUInt32BE(MAGIC_RAW, 0);
    header.writeUInt32BE(input.length, 4);
    header.writeUInt32BE(crc32(input), 8);
    return Buffer.concat([header, input]);
}

/* ── File-level Decompress ── */

function decompressFile(input) {
    if (input.length < 12) return null;

    const magic = input.readUInt32BE(0);

    // Handle raw stored files
    if (magic === MAGIC_RAW) {
        return decompressRaw(input);
    }

    if (magic !== MAGIC) return null;
    if (input.length < 22) return null;

    let pos = 4;
    const version = input[pos++];
    if (version > VERSION) return null; // Accept v1, v2, v3

    const profile = input[pos++];
    const blockSize = input.readUInt32BE(pos); pos += 4;
    const origSize = input.readUInt32BE(pos); pos += 4;
    const blockCount = input.readUInt32BE(pos); pos += 4;

    const storedCrc = input.readUInt32BE(pos); pos += 4;
    const calcCrc = crc32(input.subarray(0, 18));
    if (calcCrc !== storedCrc) return null;

    const output = Buffer.alloc(origSize);
    let outOffset = 0;

    for (let b = 0; b < blockCount; b++) {
        if (pos + 13 > input.length) return null;

        const blkOrig = input.readUInt32BE(pos);
        const blkComp = input.readUInt32BE(pos + 4);
        const blkCrc = input.readUInt32BE(pos + 8);
        const blkProfile = input[pos + 12];
        pos += 13;

        if (pos + blkComp > input.length) return null;

        let decoded;
        if (blkProfile === 0xFF) {
            decoded = input.subarray(pos, pos + blkOrig);
        } else {
            decoded = decompressBlock(input.subarray(pos, pos + blkComp), blkProfile, blkOrig);
        }

        if (decoded.length !== blkOrig) return null;
        if (crc32(decoded) !== blkCrc) return null;

        decoded.copy(output, outOffset);
        pos += blkComp;
        outOffset += blkOrig;
    }

    return output;
}

function decompressRaw(input) {
    if (input.length < 12) return null;
    const origSize = input.readUInt32BE(4);
    const storedCrc = input.readUInt32BE(8);

    if (input.length < 12 + origSize) return null;

    const data = input.subarray(12, 12 + origSize);
    if (crc32(data) !== storedCrc) return null;

    return Buffer.from(data);
}

module.exports = {
    PROFILE, PROFILE_NAMES, DEFAULT_BLOCK_SIZE,
    profileName,
    compressBlock, decompressBlock,
    compressFile, decompressFile
};

File: Byte_Compressor/src/core/block_codec.c
/*
 * ByteCompressor - Block Codec Implementation
 */

#include "block_codec.h"
#include "preprocessor.h"
#include "context_model.h"
#include "ans_codec.h"
#include "../utils/crc32.h"
#include <stdlib.h>
#include <string.h>

const char *bc_profile_name(BCProfile profile) {
    switch (profile) {
        case BC_PROFILE_GENERIC:   return "GENERIC (BWT+MTF+RLE+ANS)";
        case BC_PROFILE_TELEMETRY: return "TELEMETRY (Delta+RLE+ANS)";
        case BC_PROFILE_IMAGE:     return "IMAGE (Delta+BWT+MTF+ANS)";
        case BC_PROFILE_RAW_ANS:   return "RAW_ANS (ANS only)";
        default:                   return "UNKNOWN";
    }
}

BCConfig bc_default_config(void) {
    BCConfig cfg;
    cfg.profile = BC_PROFILE_GENERIC;
    cfg.block_size = BC_DEFAULT_BLOCK_SIZE;
    return cfg;
}

/* ── Single Block Compression ── */

size_t bc_compress_block(const uint8_t *input, size_t input_size,
                         uint8_t *output, size_t output_capacity,
                         BCProfile profile) {
    if (input_size == 0) return 0;

    /* Allocate workspace for transforms */
    size_t work_size = input_size * 4 + 1024;
    uint8_t *work1 = (uint8_t *)malloc(work_size);
    uint8_t *work2 = (uint8_t *)malloc(work_size);
    int32_t *bwt_workspace = NULL;
    if (!work1 || !work2) {
        free(work1); free(work2);
        return 0;
    }

    const uint8_t *current = input;
    size_t current_size = input_size;
    int32_t bwt_primary = 0;

    switch (profile) {
        case BC_PROFILE_GENERIC:
            /* BWT → MTF → RLE → ANS */
            bwt_workspace = (int32_t *)malloc(input_size * sizeof(int32_t));
            if (!bwt_workspace) { free(work1); free(work2); return 0; }

            bwt_primary = bwt_forward(current, work1, current_size, bwt_workspace);
            free(bwt_workspace);

            mtf_encode(work1, work2, current_size);
            current_size = rle_encode(work2, current_size, work1, work_size);
            if (current_size == 0) { free(work1); free(work2); return 0; }
            current = work1;

            /* Store BWT primary index in first 4 bytes of output */
            if (output_capacity < 4) { free(work1); free(work2); return 0; }
            output[0] = (uint8_t)(bwt_primary >> 24);
            output[1] = (uint8_t)(bwt_primary >> 16);
            output[2] = (uint8_t)(bwt_primary >> 8);
            output[3] = (uint8_t)(bwt_primary);

            {
                size_t comp_size = cm_compress(current, current_size,
                                               output + 4, output_capacity - 4);
                free(work1); free(work2);
                return (comp_size > 0) ? comp_size + 4 : 0;
            }

        case BC_PROFILE_TELEMETRY:
            /* Delta → RLE → ANS */
            delta_encode(current, work1, current_size);
            current_size = rle_encode(work1, current_size, work2, work_size);
            if (current_size == 0) { free(work1); free(work2); return 0; }

            {
                size_t comp_size = cm_compress(work2, current_size,
                                               output, output_capacity);
                free(work1); free(work2);
                return comp_size;
            }

        case BC_PROFILE_IMAGE:
            /* Delta → BWT → MTF → ANS */
            delta_encode(current, work1, current_size);

            bwt_workspace = (int32_t *)malloc(current_size * sizeof(int32_t));
            if (!bwt_workspace) { free(work1); free(work2); return 0; }

            bwt_primary = bwt_forward(work1, work2, current_size, bwt_workspace);
            free(bwt_workspace);

            mtf_encode(work2, work1, current_size);
            current = work1;

            if (output_capacity < 4) { free(work1); free(work2); return 0; }
            output[0] = (uint8_t)(bwt_primary >> 24);
            output[1] = (uint8_t)(bwt_primary >> 16);
            output[2] = (uint8_t)(bwt_primary >> 8);
            output[3] = (uint8_t)(bwt_primary);

            {
                size_t comp_size = cm_compress(current, current_size,
                                               output + 4, output_capacity - 4);
                free(work1); free(work2);
                return (comp_size > 0) ? comp_size + 4 : 0;
            }

        case BC_PROFILE_RAW_ANS:
            /* ANS only */
            {
                size_t comp_size = cm_compress(current, current_size,
                                               output, output_capacity);
                free(work1); free(work2);
                return comp_size;
            }

        default:
            free(work1); free(work2);
            return 0;
    }
}

/* ── Single Block Decompression ── */

size_t bc_decompress_block(const uint8_t *input, size_t input_size,
                           uint8_t *output, size_t output_capacity,
                           BCProfile profile) {
    if (input_size == 0) return 0;

    size_t work_size = output_capacity * 4 + 1024;
    uint8_t *work1 = (uint8_t *)malloc(work_size);
    uint8_t *work2 = (uint8_t *)malloc(work_size);
    if (!work1 || !work2) {
        free(work1); free(work2);
        return 0;
    }

    switch (profile) {
        case BC_PROFILE_GENERIC: {
            /* ANS → RLE⁻¹ → MTF⁻¹ → BWT⁻¹ */
            if (input_size < 4) { free(work1); free(work2); return 0; }
            int32_t bwt_primary = ((int32_t)input[0] << 24) | ((int32_t)input[1] << 16) |
                                  ((int32_t)input[2] << 8)  | (int32_t)input[3];

            size_t rle_size = cm_decompress(input + 4, input_size - 4, work1, work_size);
            if (rle_size == 0) { free(work1); free(work2); return 0; }

            size_t mtf_size = rle_decode(work1, rle_size, work2, work_size);
            if (mtf_size == 0) { free(work1); free(work2); return 0; }

            mtf_decode(work2, work1, mtf_size);
            bwt_inverse(work1, output, mtf_size, bwt_primary);
            free(work1); free(work2);
            return mtf_size;
        }

        case BC_PROFILE_TELEMETRY: {
            /* ANS → RLE⁻¹ → Delta⁻¹ */
            size_t rle_size = cm_decompress(input, input_size, work1, work_size);
            if (rle_size == 0) { free(work1); free(work2); return 0; }

            size_t delta_size = rle_decode(work1, rle_size, work2, work_size);
            if (delta_size == 0) { free(work1); free(work2); return 0; }

            delta_decode(work2, output, delta_size);
            free(work1); free(work2);
            return delta_size;
        }

        case BC_PROFILE_IMAGE: {
            /* ANS → MTF⁻¹ → BWT⁻¹ → Delta⁻¹ */
            if (input_size < 4) { free(work1); free(work2); return 0; }
            int32_t bwt_primary = ((int32_t)input[0] << 24) | ((int32_t)input[1] << 16) |
                                  ((int32_t)input[2] << 8)  | (int32_t)input[3];

            size_t mtf_size = cm_decompress(input + 4, input_size - 4, work1, work_size);
            if (mtf_size == 0) { free(work1); free(work2); return 0; }

            mtf_decode(work1, work2, mtf_size);
            bwt_inverse(work2, work1, mtf_size, bwt_primary);
            delta_decode(work1, output, mtf_size);
            free(work1); free(work2);
            return mtf_size;
        }

        case BC_PROFILE_RAW_ANS: {
            size_t dec_size = cm_decompress(input, input_size, output, output_capacity);
            free(work1); free(work2);
            return dec_size;
        }

        default:
            free(work1); free(work2);
            return 0;
    }
}

/* ── File-level Compress ── */

static void write_u32(uint8_t *buf, uint32_t val) {
    buf[0] = (uint8_t)(val >> 24);
    buf[1] = (uint8_t)(val >> 16);
    buf[2] = (uint8_t)(val >> 8);
    buf[3] = (uint8_t)(val);
}

static uint32_t read_u32(const uint8_t *buf) {
    return ((uint32_t)buf[0] << 24) | ((uint32_t)buf[1] << 16) |
           ((uint32_t)buf[2] << 8)  | (uint32_t)buf[3];
}

size_t bc_compress(const uint8_t *input, size_t input_size,
                   uint8_t *output, size_t output_capacity,
                   const BCConfig *config) {
    BCConfig cfg = config ? *config : bc_default_config();
    uint32_t block_size = cfg.block_size;
    uint32_t block_count = (uint32_t)((input_size + block_size - 1) / block_size);

    /* Write file header (18 bytes) */
    size_t pos = 0;
    size_t header_start = pos;
    if (pos + 18 > output_capacity) return 0;

    write_u32(output + pos, BC_MAGIC);       pos += 4;
    output[pos++] = BC_VERSION;
    output[pos++] = (uint8_t)cfg.profile;
    write_u32(output + pos, block_size);     pos += 4;
    write_u32(output + pos, (uint32_t)input_size); pos += 4;
    write_u32(output + pos, block_count);    pos += 4;

    /* Placeholder for header CRC — fill later */
    size_t crc_pos = pos;
    write_u32(output + pos, 0);              pos += 4;

    /* CRC of header (excluding the CRC field itself) */
    uint32_t hdr_crc = crc32_compute(output + header_start, crc_pos - header_start);
    write_u32(output + crc_pos, hdr_crc);

    /* Compress each block */
    size_t in_offset = 0;
    for (uint32_t b = 0; b < block_count; b++) {
        size_t chunk = input_size - in_offset;
        if (chunk > block_size) chunk = block_size;

        /* Reserve space for block header (13 bytes) */
        if (pos + 13 > output_capacity) return 0;
        size_t bh_pos = pos;
        pos += 13;

        /* Compress block */
        size_t comp_size = bc_compress_block(input + in_offset, chunk,
                                              output + pos, output_capacity - pos,
                                              cfg.profile);

        /* If compression fails or expands, store raw */
        if (comp_size == 0 || comp_size >= chunk) {
            if (pos + chunk > output_capacity) return 0;
            memcpy(output + pos, input + in_offset, chunk);
            comp_size = chunk;
            /* Mark as raw by using profile 0xFF */
            output[bh_pos + 12] = 0xFF;
        } else {
            output[bh_pos + 12] = (uint8_t)cfg.profile;
        }

        /* Write block header */
        uint32_t blk_crc = crc32_compute(input + in_offset, chunk);
        write_u32(output + bh_pos + 0, (uint32_t)chunk);
        write_u32(output + bh_pos + 4, (uint32_t)comp_size);
        write_u32(output + bh_pos + 8, blk_crc);

        pos += comp_size;
        in_offset += chunk;
    }

    return pos;
}

/* ── File-level Decompress ── */

size_t bc_decompress(const uint8_t *input, size_t input_size,
                     uint8_t *output, size_t output_capacity) {
    if (input_size < 22) return 0;

    size_t pos = 0;

    /* Read and verify file header */
    uint32_t magic = read_u32(input + pos); pos += 4;
    if (magic != BC_MAGIC) return 0;

    uint8_t version = input[pos++];
    if (version != BC_VERSION) return 0;

    uint8_t profile = input[pos++];
    uint32_t block_size = read_u32(input + pos); pos += 4;
    uint32_t orig_size = read_u32(input + pos);  pos += 4;
    uint32_t block_count = read_u32(input + pos); pos += 4;

    uint32_t stored_crc = read_u32(input + pos); pos += 4;

    /* Verify header CRC */
    uint32_t calc_crc = crc32_compute(input, pos - 4);
    if (calc_crc != stored_crc) return 0;

    if (orig_size > output_capacity) return 0;

    /* Decompress each block */
    size_t out_offset = 0;
    for (uint32_t b = 0; b < block_count; b++) {
        if (pos + 13 > input_size) return 0;

        uint32_t blk_orig = read_u32(input + pos + 0);
        uint32_t blk_comp = read_u32(input + pos + 4);
        uint32_t blk_crc  = read_u32(input + pos + 8);
        uint8_t  blk_prof = input[pos + 12];
        pos += 13;

        if (pos + blk_comp > input_size) return 0;
        if (out_offset + blk_orig > output_capacity) return 0;

        size_t dec_size;
        if (blk_prof == 0xFF) {
            /* Raw block */
            memcpy(output + out_offset, input + pos, blk_orig);
            dec_size = blk_orig;
        } else {
            dec_size = bc_decompress_block(input + pos, blk_comp,
                                            output + out_offset, blk_orig,
                                            (BCProfile)blk_prof);
        }

        /* Verify block CRC */
        if (dec_size != blk_orig) return 0;
        uint32_t check_crc = crc32_compute(output + out_offset, dec_size);
        if (check_crc != blk_crc) return 0;

        pos += blk_comp;
        out_offset += dec_size;
    }

    return out_offset;
}
